File: backend/src/routers/revisao.py
from fastapi import APIRouter, Depends, Query, status, Response
from ..schemas.revisao import Revisao, RevisaoUpdate
from ..schemas.pagination import Page
from ..models.user import User as UserModel
from ..security import get_current_user
from ..services.revisao import RevisaoService
from datetime import date

router = APIRouter(prefix="/revisoes", tags=["revisoes"])

@router.get("/", response_model=Page[Revisao])
async def get_revisoes(
    user: UserModel = Depends(get_current_user), 
    service: RevisaoService = Depends(),
    page: int = Query(1, ge=1, description="Número da página"),
    size: int = Query(10, ge=1, le=100, description="Tamanho da página")
):
    return await service.get_revisoes_for_user(user.ID, page, size)

@router.put("/{revisao_id}", response_model=Revisao)
async def update_revisao(
    revisao_id: int,
    revisao: RevisaoUpdate,
    user: UserModel = Depends(get_current_user),
    service: RevisaoService = Depends(),
):
    return await service.update_revisao(revisao_id, revisao, user.ID)

@router.put("/{revisao_id}/concluir", response_model=Revisao)
async def concluir_revisao(revisao_id: int, revisao: RevisaoUpdate, user: UserModel = Depends(get_current_user), service: RevisaoService = Depends()):
    return await service.concluir_revisao(revisao_id, revisao, user.ID)

@router.get("/cronograma", response_model=Page[Revisao])
async def get_cronograma(
    user: UserModel = Depends(get_current_user), 
    service: RevisaoService = Depends(),
    page: int = Query(1, ge=1, description="Número da página"),
    size: int = Query(10, ge=1, le=100, description="Tamanho da página")
):
    return await service.get_cronograma(user.ID, page, size)

@router.put("/{revisao_id}/reagendar", response_model=Revisao)
async def reagendar_revisao(revisao_id: int, nova_data: date, user: UserModel = Depends(get_current_user), service: RevisaoService = Depends()):
    return await service.reagendar_revisao(revisao_id, nova_data, user.ID)

@router.delete("/{revisao_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_revisao(
    revisao_id: int,
    user: UserModel = Depends(get_current_user),
    service: RevisaoService = Depends(),
):
    await service.delete_revisao(revisao_id, user.ID)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/revisao.ts
import api from './apiConnect'
import type { PageParams, Revisao } from '@/utils/apiTypes'

const resourceUrl = '/revisoes'


async function getRevisoes(p?: PageParams) {
  try {
    const response = await api.get(resourceUrl, {
     params: {
       page: p?.page,
       size: p?.size
     }
   })
    return response.data
  } catch (error: any) {
    const errorMessage = error.response?.data?.detail || 'Erro ao buscar as revisões. Por favor, tente novamente.'
    throw new Error(errorMessage)
  }
}


async function updateRevisao(id: number, revisao: Revisao) {
  try {
    const response = await api.put(`${resourceUrl}/${id}`, revisao)
    return response.data
  } catch (error: any) {
    const errorMessage = error.response?.data?.detail || 'Erro ao atualizar a revisão. Por favor, tente novamente.'
    throw new Error(errorMessage)
  }
}


async function concluirRevisao(id: number, revisao: Revisao) {
  try {
    const response = await api.put(`${resourceUrl}/${id}/concluir`, revisao)
    return response.data
  } catch (error: any) {
    const errorMessage = error.response?.data?.detail || 'Erro ao concluir a revisão. Por favor, tente novamente.'
    throw new Error(errorMessage)
  }
}

async function getCronograma(p?: PageParams) {
  try {
    const response = await api.get(`${resourceUrl}/cronograma`, {
      params: {
        page: p?.page,
        size: p?.size
      }
    })
    return response.data
  } catch (error: any) {
    const errorMessage = error.response?.data?.detail || 'Erro ao buscar o cronograma. Por favor, tente novamente.'
    throw new Error(errorMessage)
  }
}

async function reagendarRevisao(id: number, nova_data: string) {
  try {
    const response = await api.put(`${resourceUrl}/${id}/reagendar`, null, {
      params: {
        nova_data: nova_data
      }
    });
    return response.data;
  } catch (error: any) {
    const errorMessage = error.response?.data?.detail || 'Erro ao reagendar a revisão. Por favor, tente novamente.';
    throw new Error(errorMessage);
  }
}

async function deleteRevisao(id: number) {
  try {
    const response = await api.delete(`${resourceUrl}/${id}`)
    return response.data
  } catch (error: any) {
    const errorMessage = error.response?.data?.detail || 'Erro ao excluir a revisão. Por favor, tente novamente.'
    throw new Error(errorMessage)
  }
}

export {
  getRevisoes,
  updateRevisao,
  concluirRevisao,
  getCronograma,
  reagendarRevisao,
  deleteRevisao
}
